test/pkg/environment/azure/byok.go

```diff
@@ -0,0 +1,262 @@
+/*
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v5"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/keyvault/armkeyvault"
+	"github.com/Azure/azure-sdk-for-go/sdk/security/keyvault/azkeys"
+	. "github.com/onsi/gomega"
+	"github.com/samber/lo"
+)
+
+const (
+	cryptoServiceEncryptionUserRole = "e147488a-f6f5-4113-8e2d-b22465e65bf6"
+	readerRole                      = "acdd72a7-3385-48ef-bd42-f606fba81ae7"
+	keyVaultAdminRole               = "00482a5a-887f-4fb3-b363-3b7fe8e74483"
+	keyVaultCryptoOfficer           = "14b46e9e-c2b7-41b4-b07b-48a6ebf60603"
+)
+
+// CreateKeyVaultAndDiskEncryptionSet creates a key vault and disk encryption set for BYOK testing
+//
+// This function performs the following operations in sequence:
+//
+// 1. API Calls:
+//   - Creates an Azure Key Vault with RBAC authorization enabled
+//   - Creates a key in the Key Vault for encryption
+//   - Creates a Disk Encryption Set (DES) referencing the Key Vault key
+//
+// 2. RBAC Assignments:
+//   - Assigns Key Vault Crypto Officer and Administrator roles to the Karpenter workload identity
+//   - Assigns Key Vault Crypto Officer and Administrator roles to the test user (for key creation)
+//   - Assigns Reader role to the Karpenter workload identity for DES access (so it can create VMs)
+//   - Assigns Key Vault Crypto Service Encryption User role to the DES managed identity (so DES can access keys)
+//
+// 3. Retry Logic:
+//   - Azure clients are configured with retry logic for 403 (Forbidden) errors caused by RBAC propagation delays
+//   - Azure clients also retry on 400 (Bad Request) errors for PrincipalNotFound when identities haven't replicated
+//
+// Returns the DES resource ID for use in VM disk encryption configuration.
+func (env *Environment) CreateKeyVaultAndDiskEncryptionSet(ctx context.Context) string {
```

Comment:
Is this really a primitive we want/need on env?

Feels like possibly instead of being on env it should just live in the BYOK suite and be built out of the (would need to be exported) primitive functions for making a KV and/or DES?